File: desktop/src/sources.ts
import { equivalentISBN, validISBNs } from "./isbn.ts";
import { splitManifestationStatement } from "./manifestation.ts";
import { FetchError, RequestPacer, applicationUserAgent, fetchJSON, fetchText, mapLimit } from "./network.ts";
import { sourceRecord, type ISBNInfo, type SourceRecord, type SourceResult } from "./types.ts";

const LANGUAGES: Record<string, string> = {
  id: "Indonesian",
  ind: "Indonesian",
  en: "English",
  eng: "English",
  jv: "Javanese",
  jav: "Javanese",
  ms: "Malay",
  msa: "Malay",
  may: "Malay",
};

export interface SourceAdapter {
  name: string;
  search(isbn: ISBNInfo): Promise<SourceResult>;
}

interface Pacer {
  wait(): Promise<void>;
}

type JsonObject = Record<string, unknown>;

function asObject(value: unknown): JsonObject {
  return value && typeof value === "object" && !Array.isArray(value) ? value as JsonObject : {};
}

function asArray(value: unknown): unknown[] {
  return Array.isArray(value) ? value : [];
}

export function clean(value: unknown): string {
  return String(value ?? "").replaceAll("\0", "").trim().replace(/\s+/g, " ");
}

export function unique(values: Iterable<unknown>): string[] {
  const output: string[] = [];
  const seen = new Set<string>();
  for (const raw of values) {
    const value = clean(raw);
    const marker = value.toLocaleLowerCase();
    if (value && !seen.has(marker)) {
      seen.add(marker);
      output.push(value);
    }
  }
  return output;
}

function language(value: unknown): string {
  const cleaned = clean(value);
  return LANGUAGES[cleaned.toLocaleLowerCase()] ?? cleaned;
}

function dictText(value: unknown): string {
  if (value && typeof value === "object" && !Array.isArray(value)) return clean((value as JsonObject).value);
  return clean(value);
}

function authorsFromByStatement(value: unknown): string[] {
  const statement = clean(value);
  if (!statement) return [];
  let head = statement.split(";", 1)[0];
  if (/\b(?:editor|edited|penyunting|penerjemah|translator|kata pengantar)\b/i.test(head)) return [];
  head = head.replace(/^(?:oleh|by)\s+/i, "");
  const pieces = head.split(/\s+(?:dan|and)\s+|\s*&\s*/i);
  if (pieces.length < 2) return [];
  const authors = pieces.map(piece => clean(piece).replace(/^(?:Dr\.?|Prof\.?)\s+/i, "")).filter(Boolean);
  return authors.length >= 2 && authors.length <= 6 ? authors : [];
}

function recordMatches(recordISBNs: string[], wanted: ISBNInfo): boolean {
  return recordISBNs.some(value => equivalentISBN(value, wanted.canonical));
}

function decodeXML(value: string): string {
  const stripped = value.trim().replace(/^<!\[CDATA\[([\s\S]*)\]\]>$/, "$1");
  return stripped
    .replace(/&#x([0-9a-f]+);/gi, (_match, digits) => String.fromCodePoint(Number.parseInt(digits, 16)))
    .replace(/&#(\d+);/g, (_match, digits) => String.fromCodePoint(Number.parseInt(digits, 10)))
    .replaceAll("&amp;", "&")
    .replaceAll("&lt;", "<")
    .replaceAll("&gt;", ">")
    .replaceAll("&quot;", "\"")
    .replaceAll("&apos;", "'");
}

function rssLinks(xml: string): string[] {
  if (!/<(?:rss|channel)\b/i.test(xml)) throw new FetchError("", "Source returned malformed XML");
  const output: string[] = [];
  for (const item of xml.match(/<item\b[\s\S]*?<\/item>/gi) ?? []) {
    const match = item.match(/<link\b[^>]*>([\s\S]*?)<\/link>/i);
    const link = match ? clean(decodeXML(match[1])) : "";
    if (link) output.push(link);
  }
  return output;
}

export class IndonesiaOneSearch implements SourceAdapter {
  readonly name = "Indonesia OneSearch";
  readonly baseURL = "https://onesearch.id";
  readonly maxRecords: number;
  readonly pacer: Pacer;

  constructor(maxRecords = 8, pacer: Pacer = new RequestPacer(1_000)) {
    this.maxRecords = maxRecords;
    this.pacer = pacer;
  }

  async search(isbn: ISBNInfo): Promise<SourceResult> {
    const links: string[] = [];
    const errors: string[] = [];
    const formResults = await Promise.all(isbn.searchForms.map(form => this.searchForm(form)));
    for (const result of formResults) {
      errors.push(...result.errors);
      for (const link of result.links) if (!links.includes(link)) links.push(link);
    }

    const selected = links.slice(0, this.maxRecords);
    const records: SourceRecord[] = [];
    if (selected.length) {
      const fetched = await mapLimit(selected, 8, async link => {
        try {
          return { link, record: await this.fetchRecord(link, isbn), error: "" };
        } catch (error) {
          return { link, record: null, error: error instanceof Error ? error.message : String(error) };
        }
      });
      for (const result of fetched) {
        if (result.error) errors.push(result.error);
        if (result.record) records.push(result.record);
      }
      const order = new Map(selected.map((link, index) => [link, index]));
      records.sort((left, right) => (order.get(left.source_url) ?? selected.length) - (order.get(right.source_url) ?? selected.length));
    }

    if (records.length) {
      return {
        records,
        status: {
          source: this.name,
          ok: true,
          records: records.length,
          message: errors.length ? `${errors.length} catalogue request(s) failed; usable records were retained.` : "",
        },
      };
    }
    if (errors.length) {
      return { records: [], status: { source: this.name, ok: false, records: 0, message: errors[0] } };
    }
    return {
      records: [],
      status: { source: this.name, ok: true, records: 0, message: "No matching catalogue record" },
    };
  }

  private async searchForm(form: string): Promise<{ links: string[]; errors: string[] }> {
    const errors: string[] = [];
    for (const searchType of ["ISN", "AllFields"]) {
      const query = new URLSearchParams({ lookfor: form, type: searchType, view: "rss", limit: "50" });
      const url = `${this.baseURL}/Search/Results?${query}`;
      try {
        await this.pacer.wait();
        const xml = await fetchText(url, "application/rss+xml, application/xml");
        const links = rssLinks(xml);
        if (links.length) return { links, errors };
      } catch (error) {
        errors.push(error instanceof Error ? error.message : String(error));
      }
    }
    return { links: [], errors };
  }

  private async fetchRecord(link: string, wanted: ISBNInfo): Promise<SourceRecord | null> {
    const exportURL = `${link.replace(/\/+$/, "")}/Export?style=EndNote`;
    await this.pacer.wait();
    const text = await fetchText(exportURL, "application/x-endnote-refer, text/plain");
    const fields = IndonesiaOneSearch.parseEndNote(text);
    const isbnValues = validISBNs(fields["@"] ?? []);
    if (!recordMatches(isbnValues, wanted)) return null;

    let publisher = clean(fields.I?.[0]);
    let place = clean(fields.C?.[0]);
    let date = clean(fields.D?.[0]);
    const imprint = publisher.match(/^(.+?)\s*:\s*(.+?)\s*,\s*(\d{4})$/);
    if (imprint) {
      place = place || clean(imprint[1]);
      publisher = clean(imprint[2]);
      date = date || imprint[3];
    }

    const extents = (fields["0"] ?? [])
      .map(value => value.replace(/^Other\s*:\s*/i, "").trim())
      .filter(value => /\b(?:hlm|halaman|pages?|pg)\b/i.test(value));
    const extent = extents.sort((left, right) => right.length - left.length)[0] ?? "";
    const title = clean(fields.T?.[0]);
    if (!title) return null;
    const manifestation = splitManifestationStatement(fields["7"]?.[0]);
    return sourceRecord({
      source: this.name,
      source_id: link.replace(/\/+$/, "").split("/").at(-1) ?? link,
      source_url: link,
      title,
      authors: unique(fields.A ?? []),
      editors: unique(fields.E ?? []),
      publisher,
      place,
      date,
      edition: manifestation.edition,
      printing: manifestation.printing,
      extent,
      languages: unique((fields.G ?? []).map(language)),
      isbns: isbnValues,
      subjects: unique(fields.K ?? []),
      abstract: clean((fields.X ?? []).join(" ")),
      notes: unique(fields.N ?? []),
    });
  }

  static parseEndNote(text: string): Record<string, string[]> {
    const fields: Record<string, string[]> = {};
    let current: string | null = null;
    for (const rawLine of text.replaceAll("\r\n", "\n").replaceAll("\r", "\n").split("\n")) {
      const match = rawLine.match(/^%(.?)\s+(.*)$/);
      if (match) {
        current = match[1];
        (fields[current] ??= []).push(clean(match[2]));
      } else if (current && /^[ \t]/.test(rawLine) && fields[current]?.length) {
        const index = fields[current].length - 1;
        fields[current][index] = clean(`${fields[current][index]} ${rawLine}`);
      }
    }
    return fields;
  }
}

export class OpenLibrary implements SourceAdapter {
  readonly name = "Open Library";
  readonly baseURL = "https://openlibrary.org";
  readonly contact: string;
  readonly userAgent: string;
  readonly pacer: Pacer;

  constructor(contact = process.env.OPEN_LIBRARY_CONTACT ?? "", pacer?: Pacer) {
    this.contact = String(contact).trim();
    this.userAgent = applicationUserAgent(this.contact);
    this.pacer = pacer ?? new RequestPacer(this.contact ? 334 : 1_000);
  }

  async search(isbn: ISBNInfo): Promise<SourceResult> {
    const bibkeys = isbn.searchForms.map(form => `ISBN:${form}`).join(",");
    const query = new URLSearchParams({ bibkeys, jscmd: "data", format: "json" });
    const url = `${this.baseURL}/api/books?${query}`;
    let payload: JsonObject;
    try {
      await this.pacer.wait();
      payload = await fetchJSON(url, 18_000, this.userAgent);
    } catch (error) {
      return {
        records: [],
        status: { source: this.name, ok: false, records: 0, message: error instanceof Error ? error.message : String(error) },
      };
    }

    const editions = new Map<string, JsonObject>();
    const matchedForms = new Map<string, string[]>();
    for (const [bibkey, rawData] of Object.entries(payload)) {
      const data = asObject(rawData);
      if (!Object.keys(data).length) continue;
      const key = clean(data.key) || clean(data.url) || bibkey;
      editions.set(key, data);
      const form = bibkey.split(":", 2).at(-1) ?? bibkey;
      const values = matchedForms.get(key) ?? [];
      values.push(form);
      matchedForms.set(key, values);
    }

    const records: SourceRecord[] = [];
    for (const [key, data] of editions) {
      const record = this.recordFromAPIBook(key, data, matchedForms.get(key) ?? [], isbn);
      if (record) records.push(record);
    }
    if (records.length) return { records, status: { source: this.name, ok: true, records: records.length, message: "" } };
    return {
      records: [],
      status: { source: this.name, ok: true, records: 0, message: "No matching edition record" },
    };
  }

  private recordFromAPIBook(key: string, data: JsonObject, queryForms: string[], wanted: ISBNInfo): SourceRecord | null {
    const identifiers = asObject(data.identifiers);
    const isbnValues = validISBNs([
      ...asArray(identifiers.isbn_13),
      ...asArray(identifiers.isbn_10),
      ...queryForms,
    ]);
    if (!recordMatches(isbnValues, wanted)) return null;

    const authors = asArray(data.authors)
      .map(author => clean(asObject(author).name))
      .filter(Boolean);
    const notes: string[] = [];
    const byStatement = clean(data.by_statement);
    if (byStatement) {
      notes.push(`Statement of responsibility: ${byStatement}`);
      authors.push(...authorsFromByStatement(byStatement));
    }
    const contributions = unique(asArray(data.contributions));
    if (contributions.length) notes.push(`Other contributions listed by source: ${contributions.join("; ")}`);
    const sourceNotes = dictText(data.notes);
    if (sourceNotes) notes.push(sourceNotes);

    const languages = asArray(data.languages).map(value => {
      const item = asObject(value);
      const keyValue = Object.keys(item).length ? clean(item.key || item.name) : clean(value);
      return language(keyValue.split("/").at(-1));
    });
    const sourceIdentifiers: Record<string, string[]> = {};
    for (const [field, label] of [["lccn", "LCCN"], ["oclc", "OCLC"]] as const) {
      const values = unique(asArray(identifiers[field]));
      if (values.length) sourceIdentifiers[label] = values;
    }

    const namedValues = (value: unknown): string[] => asArray(value).map(item => {
      const objectValue = asObject(item);
      return Object.keys(objectValue).length ? clean(objectValue.name) : clean(item);
    });
    let sourceURL = clean(data.url);
    if (sourceURL.startsWith("http://")) sourceURL = `https://${sourceURL.slice("http://".length)}`;
    if (!sourceURL) sourceURL = `${this.baseURL}${key}`;
    const title = clean(data.title);
    if (!title) return null;
    const manifestation = splitManifestationStatement(data.edition_name);
    return sourceRecord({
      source: this.name,
      source_id: key.split("/").at(-1) ?? key,
      source_url: sourceURL,
      title,
      subtitle: clean(data.subtitle),
      authors: unique(authors),
      publisher: namedValues(data.publishers).find(Boolean) ?? "",
      place: namedValues(data.publish_places).find(Boolean) ?? "",
      date: clean(data.publish_date),
      edition: manifestation.edition,
      printing: manifestation.printing,
      num_pages: clean(data.number_of_pages),
      extent: clean(data.pagination),
      languages: unique(languages),
      isbns: isbnValues,
      subjects: unique(namedValues(data.subjects)),
      abstract: dictText(data.description),
      notes: unique(notes),
      identifiers: sourceIdentifiers,
    });
  }
}

export class GoogleBooks implements SourceAdapter {
  // Dormant adapter: use only in a separate, non-intermixed compliant view.
  readonly name = "Google Books";
  readonly apiURL = "https://www.googleapis.com/books/v1/volumes";
  readonly apiKey: string;

  constructor(apiKey = process.env.GOOGLE_BOOKS_API_KEY ?? "") {
    this.apiKey = apiKey;
  }

  async search(isbn: ISBNInfo): Promise<SourceResult> {
    const errors: string[] = [];
    const responses = await Promise.all(isbn.searchForms.map(async form => {
      const params: Record<string, string> = { q: `isbn:${form}`, maxResults: "40", projection: "full" };
      if (this.apiKey) params.key = this.apiKey;
      const url = `${this.apiURL}?${new URLSearchParams(params)}`;
      try {
        return await fetchJSON(url);
      } catch (error) {
        errors.push(error instanceof Error ? error.message : String(error));
        return {};
      }
    }));

    const items = new Map<string, JsonObject>();
    for (const response of responses) {
      for (const rawItem of asArray(response.items)) {
        const item = asObject(rawItem);
        const id = clean(item.id);
        if (id) items.set(id, item);
      }
    }

    const records: SourceRecord[] = [];
    for (const [id, item] of items) {
      const info = asObject(item.volumeInfo);
      const identifierValues = asArray(info.industryIdentifiers).map(value => clean(asObject(value).identifier));
      const isbnValues = validISBNs(identifierValues);
      if (!recordMatches(isbnValues, isbn)) continue;
      const title = clean(info.title);
      if (!title) continue;
      records.push(sourceRecord({
        source: this.name,
        source_id: id,
        source_url: `https://books.google.com/books?id=${encodeURIComponent(id)}`,
        title,
        subtitle: clean(info.subtitle),
        authors: unique(asArray(info.authors)),
        publisher: clean(info.publisher),
        date: clean(info.publishedDate),
        num_pages: clean(info.pageCount),
        languages: unique([language(info.language)]),
        isbns: isbnValues,
        subjects: unique(asArray(info.categories)),
        abstract: clean(info.description),
      }));
    }

    if (records.length) {
      return {
        records,
        status: {
          source: this.name,
          ok: true,
          records: records.length,
          message: errors.length ? "Some Google Books requests failed; verified matches were retained." : "",
        },
      };
    }
    if (errors.length) return { records: [], status: { source: this.name, ok: false, records: 0, message: errors[0] } };
    return {
      records: [],
      status: { source: this.name, ok: true, records: 0, message: "No matching volume record" },
    };
  }
}

export function defaultSources(environment: Record<string, string | undefined> = process.env): SourceAdapter[] {
  const sources: SourceAdapter[] = [new OpenLibrary(environment.OPEN_LIBRARY_CONTACT ?? "")];
  if (["1", "true", "yes", "on"].includes(String(environment.ISBN_ZOTERO_ENABLE_ONESEARCH ?? "").trim().toLowerCase())) {
    sources.unshift(new IndonesiaOneSearch());
  }
  return sources;
}

File: python/isbn_zotero/sources.py
from __future__ import annotations

import os
import re
import urllib.parse
import xml.etree.ElementTree as ET
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Iterable, Mapping

from .isbn import ISBNInfo, ISBNValidationError, equivalent_isbn, parse_isbn, valid_isbns
from .manifestation import split_manifestation_statement
from .models import SourceRecord, SourceStatus
from .network import FetchError, HTTPClient, RequestPacer, application_user_agent


LANGUAGES = {
    "id": "Indonesian",
    "ind": "Indonesian",
    "en": "English",
    "eng": "English",
    "jv": "Javanese",
    "jav": "Javanese",
    "ms": "Malay",
    "msa": "Malay",
    "may": "Malay",
}


def _clean(value: object) -> str:
    return " ".join(str(value or "").replace("\x00", "").split())


def _unique(values: Iterable[str]) -> list[str]:
    output: list[str] = []
    seen: set[str] = set()
    for raw in values:
        value = _clean(raw)
        marker = value.casefold()
        if value and marker not in seen:
            seen.add(marker)
            output.append(value)
    return output


def _language(value: str) -> str:
    value = _clean(value)
    return LANGUAGES.get(value.casefold(), value)


def _dict_text(value: object) -> str:
    if isinstance(value, dict):
        return _clean(value.get("value", ""))
    return _clean(value)


def _authors_from_by_statement(value: str) -> list[str]:
    statement = _clean(value)
    if not statement:
        return []
    head = statement.split(";", 1)[0]
    if re.search(r"\b(?:editor|edited|penyunting|penerjemah|translator|kata pengantar)\b", head, flags=re.I):
        return []
    head = re.sub(r"^(?:oleh|by)\s+", "", head, flags=re.I)
    pieces = re.split(r"\s+(?:dan|and)\s+|\s*&\s*", head, flags=re.I)
    if len(pieces) < 2:
        return []
    authors = []
    for piece in pieces:
        name = re.sub(r"^(?:Dr\.?|Prof\.?)\s+", "", _clean(piece), flags=re.I)
        if name:
            authors.append(name)
    return authors if 2 <= len(authors) <= 6 else []


def _record_matches(record_isbns: list[str], wanted: ISBNInfo) -> bool:
    return any(equivalent_isbn(value, wanted.canonical) for value in record_isbns)


class IndonesiaOneSearch:
    name = "Indonesia OneSearch"
    base_url = "https://onesearch.id"

    def __init__(
        self,
        client: HTTPClient | None = None,
        max_records: int = 8,
        pacer: RequestPacer | None = None,
    ):
        self.client = client or HTTPClient()
        self.max_records = max_records
        self.pacer = pacer or RequestPacer(1.0)

    def search(self, isbn: ISBNInfo) -> tuple[list[SourceRecord], SourceStatus]:
        links: list[str] = []
        errors: list[str] = []
        with ThreadPoolExecutor(max_workers=len(isbn.search_forms)) as executor:
            form_results = list(executor.map(self._search_form, isbn.search_forms))
        for found, form_errors in form_results:
            errors.extend(form_errors)
            for link in found:
                if link not in links:
                    links.append(link)

        links = links[: self.max_records]
        records: list[SourceRecord] = []
        if links:
            worker_count = min(8, len(links))
            with ThreadPoolExecutor(max_workers=worker_count) as executor:
                futures = {executor.submit(self._fetch_record, link, isbn): link for link in links}
                for future in as_completed(futures):
                    try:
                        record = future.result()
                    except FetchError as error:
                        errors.append(str(error))
                        continue
                    if record:
                        records.append(record)
            order = {link: index for index, link in enumerate(links)}
            records.sort(key=lambda item: order.get(item.source_url, len(order)))

        if records:
            message = ""
            if errors:
                message = f"{len(errors)} catalogue request(s) failed; usable records were retained."
            return records, SourceStatus(self.name, True, len(records), message)
        if errors:
            return [], SourceStatus(self.name, False, 0, errors[0])
        return [], SourceStatus(self.name, True, 0, "No matching catalogue record")

    def _search_form(self, form: str) -> tuple[list[str], list[str]]:
        form_errors: list[str] = []
        found: list[str] = []
        for search_type in ("ISN", "AllFields"):
            query = urllib.parse.urlencode(
                {
                    "lookfor": form,
                    "type": search_type,
                    "view": "rss",
                    "limit": "50",
                }
            )
            url = f"{self.base_url}/Search/Results?{query}"
            try:
                self.pacer.wait()
                root = ET.fromstring(self.client.get_bytes(url, "application/rss+xml, application/xml"))
                found = [
                    _clean(item.findtext("link"))
                    for item in root.findall("./channel/item")
                    if _clean(item.findtext("link"))
                ]
            except (FetchError, ET.ParseError) as error:
                form_errors.append(str(error))
                continue
            if found:
                break
        return found, form_errors

    def _fetch_record(self, link: str, wanted: ISBNInfo) -> SourceRecord | None:
        export_url = link.rstrip("/") + "/Export?style=EndNote"
        self.pacer.wait()
        text = self.client.get_text(export_url, "application/x-endnote-refer, text/plain")
        fields = self._parse_endnote(text)
        isbn_values = valid_isbns(fields.get("@", []))
        if not _record_matches(isbn_values, wanted):
            return None

        publisher = _clean((fields.get("I") or [""])[0])
        place = _clean((fields.get("C") or [""])[0])
        date = _clean((fields.get("D") or [""])[0])
        imprint_match = re.match(r"^(.+?)\s*:\s*(.+?)\s*,\s*(\d{4})$", publisher)
        if imprint_match:
            place = place or _clean(imprint_match.group(1))
            publisher = _clean(imprint_match.group(2))
            date = date or imprint_match.group(3)

        format_values = fields.get("0", [])
        extent_candidates = []
        for value in format_values:
            cleaned = re.sub(r"^Other\s*:\s*", "", value, flags=re.I).strip()
            if re.search(r"\b(?:hlm|halaman|pages?|pg)\b", cleaned, flags=re.I):
                extent_candidates.append(cleaned)
        extent = max(extent_candidates, key=len, default="")

        record_id = link.rstrip("/").rsplit("/", 1)[-1]
        title = _clean((fields.get("T") or [""])[0])
        if not title:
            return None
        edition, printing = split_manifestation_statement((fields.get("7") or [""])[0])
        return SourceRecord(
            source=self.name,
            source_id=record_id,
            source_url=link,
            title=title,
            authors=_unique(fields.get("A", [])),
            editors=_unique(fields.get("E", [])),
            publisher=publisher,
            place=place,
            date=date,
            edition=edition,
            printing=printing,
            extent=extent,
            languages=_unique(_language(value) for value in fields.get("G", [])),
            isbns=isbn_values,
            subjects=_unique(fields.get("K", [])),
            abstract=_clean(" ".join(fields.get("X", []))),
            notes=_unique(fields.get("N", [])),
            raw={"endnote": fields, "export_url": export_url},
        )

    @staticmethod
    def _parse_endnote(text: str) -> dict[str, list[str]]:
        fields: dict[str, list[str]] = {}
        current: str | None = None
        for raw_line in text.replace("\r\n", "\n").replace("\r", "\n").split("\n"):
            match = re.match(r"^%(.?)\s+(.*)$", raw_line)
            if match:
                current = match.group(1)
                fields.setdefault(current, []).append(_clean(match.group(2)))
            elif current and raw_line.startswith((" ", "\t")) and fields.get(current):
                fields[current][-1] = _clean(fields[current][-1] + " " + raw_line)
        return fields


class OpenLibrary:
    name = "Open Library"
    base_url = "https://openlibrary.org"

    def __init__(
        self,
        client: HTTPClient | None = None,
        contact: str | None = None,
        pacer: RequestPacer | None = None,
    ):
        self.client = client or HTTPClient()
        self.contact = contact if contact is not None else os.getenv("OPEN_LIBRARY_CONTACT", "")
        self.user_agent = application_user_agent(self.contact)
        self.pacer = pacer or RequestPacer(1 / 3 if self.contact else 1.0)

    def search(self, isbn: ISBNInfo) -> tuple[list[SourceRecord], SourceStatus]:
        bibkeys = ",".join(f"ISBN:{form}" for form in isbn.search_forms)
        query = urllib.parse.urlencode({"bibkeys": bibkeys, "jscmd": "data", "format": "json"})
        url = f"{self.base_url}/api/books?{query}"
        try:
            self.pacer.wait()
            payload = self.client.get_json(url, user_agent=self.user_agent)
        except FetchError as error:
            return [], SourceStatus(self.name, False, 0, str(error))

        editions: dict[str, dict] = {}
        matched_forms: dict[str, list[str]] = {}
        for bibkey, data in payload.items():
            if not isinstance(data, dict):
                continue
            key = _clean(data.get("key")) or _clean(data.get("url")) or bibkey
            editions[key] = data
            form = bibkey.split(":", 1)[-1]
            matched_forms.setdefault(key, []).append(form)

        records: list[SourceRecord] = []
        for key, data in editions.items():
            record = self._record_from_api_book(key, data, matched_forms.get(key, []), isbn)
            if record:
                records.append(record)

        if records:
            return records, SourceStatus(self.name, True, len(records), "")
        return [], SourceStatus(self.name, True, 0, "No matching edition record")

    def _record_from_api_book(
        self,
        key: str,
        data: dict,
        query_forms: list[str],
        wanted: ISBNInfo,
    ) -> SourceRecord | None:
        source_identifiers = data.get("identifiers", {}) if isinstance(data.get("identifiers"), dict) else {}
        isbn_values = valid_isbns(
            list(source_identifiers.get("isbn_13", []))
            + list(source_identifiers.get("isbn_10", []))
            + query_forms
        )
        if not _record_matches(isbn_values, wanted):
            return None

        authors = [
            _clean(author.get("name"))
            for author in data.get("authors", [])
            if isinstance(author, dict) and _clean(author.get("name"))
        ]

        notes: list[str] = []
        by_statement = _clean(data.get("by_statement"))
        if by_statement:
            notes.append(f"Statement of responsibility: {by_statement}")
            authors.extend(_authors_from_by_statement(by_statement))
        contributions = _unique(data.get("contributions", []))
        if contributions:
            notes.append("Other contributions listed by source: " + "; ".join(contributions))
        source_notes = _dict_text(data.get("notes"))
        if source_notes:
            notes.append(source_notes)

        languages = []
        for language in data.get("languages", []):
            if isinstance(language, dict):
                key_value = _clean(language.get("key") or language.get("name"))
            else:
                key_value = _clean(language)
            languages.append(_language(key_value.rsplit("/", 1)[-1]))

        identifiers: dict[str, list[str]] = {}
        for field, label in (("lccn", "LCCN"), ("oclc", "OCLC")):
            values = _unique(source_identifiers.get(field, []))
            if values:
                identifiers[label] = values

        record_id = key.rsplit("/", 1)[-1]
        publishers = [
            _clean(item.get("name")) if isinstance(item, dict) else _clean(item)
            for item in data.get("publishers", [])
        ]
        places = [
            _clean(item.get("name")) if isinstance(item, dict) else _clean(item)
            for item in data.get("publish_places", [])
        ]
        subjects = [
            _clean(item.get("name")) if isinstance(item, dict) else _clean(item)
            for item in data.get("subjects", [])
        ]
        source_url = _clean(data.get("url"))
        if source_url.startswith("http://"):
            source_url = "https://" + source_url[len("http://") :]
        if not source_url:
            source_url = f"{self.base_url}{key}"
        edition, printing = split_manifestation_statement(data.get("edition_name"))
        return SourceRecord(
            source=self.name,
            source_id=record_id,
            source_url=source_url,
            title=_clean(data.get("title")),
            subtitle=_clean(data.get("subtitle")),
            authors=_unique(authors),
            publisher=next((value for value in publishers if value), ""),
            place=next((value for value in places if value), ""),
            date=_clean(data.get("publish_date")),
            edition=edition,
            printing=printing,
            num_pages=_clean(data.get("number_of_pages")),
            extent=_clean(data.get("pagination")),
            languages=_unique(languages),
            isbns=isbn_values,
            subjects=_unique(subjects),
            abstract=_dict_text(data.get("description")),
            notes=_unique(notes),
            identifiers=identifiers,
            raw=data,
        )


class GoogleBooks:
    """Dormant adapter: use only in a separate, non-intermixed compliant view."""

    name = "Google Books"
    api_url = "https://www.googleapis.com/books/v1/volumes"

    def __init__(self, client: HTTPClient | None = None, api_key: str | None = None):
        self.client = client or HTTPClient()
        self.api_key = api_key if api_key is not None else os.getenv("GOOGLE_BOOKS_API_KEY", "")

    def search(self, isbn: ISBNInfo) -> tuple[list[SourceRecord], SourceStatus]:
        items: dict[str, dict] = {}
        errors: list[str] = []
        for form in isbn.search_forms:
            params = {"q": f"isbn:{form}", "maxResults": "40", "projection": "full"}
            if self.api_key:
                params["key"] = self.api_key
            url = self.api_url + "?" + urllib.parse.urlencode(params)
            try:
                data = self.client.get_json(url)
            except FetchError as error:
                errors.append(str(error))
                continue
            for item in data.get("items", []) or []:
                item_id = _clean(item.get("id"))
                if item_id:
                    items[item_id] = item

        records: list[SourceRecord] = []
        for item_id, item in items.items():
            info = item.get("volumeInfo", {}) or {}
            identifier_values = [
                _clean(identifier.get("identifier"))
                for identifier in info.get("industryIdentifiers", []) or []
                if isinstance(identifier, dict)
            ]
            isbn_values = valid_isbns(identifier_values)
            if not _record_matches(isbn_values, isbn):
                continue
            title = _clean(info.get("title"))
            if not title:
                continue
            records.append(
                SourceRecord(
                    source=self.name,
                    source_id=item_id,
                    source_url=f"https://books.google.com/books?id={urllib.parse.quote(item_id)}",
                    title=title,
                    subtitle=_clean(info.get("subtitle")),
                    authors=_unique(info.get("authors", []) or []),
                    publisher=_clean(info.get("publisher")),
                    date=_clean(info.get("publishedDate")),
                    num_pages=_clean(info.get("pageCount")),
                    languages=_unique([_language(_clean(info.get("language")))]),
                    isbns=isbn_values,
                    subjects=_unique(info.get("categories", []) or []),
                    abstract=_clean(info.get("description")),
                    raw=item,
                )
            )

        if records:
            message = "" if not errors else "Some Google Books requests failed; verified matches were retained."
            return records, SourceStatus(self.name, True, len(records), message)
        if errors:
            return [], SourceStatus(self.name, False, 0, errors[0])
        return [], SourceStatus(self.name, True, 0, "No matching volume record")


DEFAULT_SOURCES = (OpenLibrary,)


def configured_source_classes(environment: Mapping[str, str] | None = None) -> tuple[type, ...]:
    values = environment if environment is not None else os.environ
    enabled = str(values.get("ISBN_ZOTERO_ENABLE_ONESEARCH", "")).strip().casefold() in {"1", "true", "yes", "on"}
    return (IndonesiaOneSearch, OpenLibrary) if enabled else DEFAULT_SOURCES
